### ISSUE TITLE
feat: better support for visual regression testing

### ISSUE THREAD

**Issue by aslushnikov** [microsoft/playwright#8161]
Playwright Test has a built-in `toMatchSnapshot()` method to power Visual Regression Testing (VRT).

However, VRT is still challenging due to variances in the host environments. There's a bunch of measures we can do right away to drastically improve experience in @playwright/test
- [ ] support for `docker` test fixture to run browsers inside docker image.
- [ ] support for `blur` in matching snapshot to counteract antialiasing

Interesting context:
- [migration from backstopjs to @playwright/test](https://github.com/microsoft/playwright/issues/7548#issuecomment-881897256)

**Comment by florianbepunkt** [comment]
I think https://github.com/americanexpress/jest-image-snapshot provides a nice suite of options for various VRT scenarios. Test scenarios vary widely, depending on the context (testing components, whole pages, text-heavy or not, etc).

Besides bluring which helps a lot with antialiasing it would be nice if multiple image comparisons (e. g. SSM) would be possible. Alternative image comparison algorithms could be left to userland, if they can be plugged into `toMatchSnapshot` via a common interface.

**Comment by aslushnikov** [comment]
> Besides bluring which helps a lot with antialiasing it would be nice if multiple image comparisons (e. g. SSM) would be possible.

@florianbepunkt What's SSM? Is it structural similarity measurement (SSIM)?

**Comment by kevinmpowell** [comment]
Solid integration with Storybook would be beneficial for the work I do. [Chromatic](https://www.chromatic.com) and [Percy](https://percy.io) do this really well.

Also a UI for reviewing the diffs would be great.

**Comment by kevinmpowell** [comment]
Slider is rarely useful for me. An onion-skin (transparency overlay) would be more useful.

**Comment by AlexNetman** [comment]
@aslushnikov Why `toMatchSnapshot()` is not available in the documentation?
It can not be found in API list.
And the article that was in 1.13  https://playwright.dev/python/docs/1.13.0/test-snapshots
is not available for 1.14 anymore.

Thanks for thinking about Visual Regression testing. Thats important!

**Comment by florianbepunkt** [comment]
On a related note: It would be great if tests could be run cross-plattform. Currently the os platform name is baked into the snapshot filename, so our CI tests sometime fail due to name miss-match. https://github.com/microsoft/playwright/issues/7575

**Comment by lo1tuma** [comment]
> support for blur in matching snapshot to counteract antialiasing

It would be nice if we could choose whether we want to apply such image filters before the snapshot is being saved or only when doing the comparison. I would prefer the first option as it keeps the diff small when creating new snapshots even of such images that change randomly / are flaky.

**Comment by ts-23** [comment]
Please allow an auto-generated filename when toMatchSnapshot has no name input, similar to how toMatchSnapshot works in Jest.

- [ ] Auto-gen filename when name not specified for toMatchSnapShot
- [ ] Set default toMatchSnapshot file extension in playwright.config.ts

E.g.

// foo.spec.ts
toMatchSnapshot() => foo.spec.ts.snap (default extension customizable in playwright.config.ts)

When you have a lot of screenshot assertions in one file, we can avoid writing a lot of filename inputs:
![image](https://user-images.githubusercontent.com/13111256/131485788-678e9baf-d416-4f17-9c75-26ef2dedfaed.png)

**Comment by sergioariveros** [comment]
Thanks for thinking on this, blur feature is something that will help us, we have something similar before with puppeter that help us to do comparisson in animated pages, in addition to that something that can be really useful is be able to ignore specific parts of the screen, specially in those parts where we have more dynamic data(videos/images)

**Comment by Doug-Bowen** [comment]
Blur would help us greatly.  Also, the slider view would be incredible as well.

**Comment by z0n** [comment]
We're also really interested in these improvements. We had to disable visual tests for now because they are randomly failing because a few pixels are off, even when increasing the threshold. Blur should help here hopefully.

**Comment by damaon** [comment]
I suggest solving biggest pain-point which is how to store this stuff in git repo so it doesn't blow up in size (to store only last snapshot). Git LFS kinda works but it's painful. Maybe something else would work better? For reference: https://github.com/americanexpress/jest-image-snapshot/issues/92

**Comment by z0n** [comment]
We're using Git LFS, what's your issue with it? Once we had it set up for everyone (we're using Mac, Windows and Linux), it worked fine. We're storing all images in the repo using Git LFS (`*.png`) so there's no work involved when adding snapshots to new tests either.

The only issue I have is comparing the image diffs in VS Code when committing new images as the old image is not shown in the diff view.

**Comment by z0n** [comment]
Hi @aslushnikov! This was pushed to the next version a few times now, could you please add this to the roadmap (if there is one?) so we can have a rough estimate on when this is coming?

I need to implement some visual tests soon™️ and it would be great if I wouldn't need another tool for that. I need to know if there will be improvements to this in 2 months or 2 years though.

**Comment by damaon** [comment]
> We're using Git LFS, what's your issue with it? 

It works for me but for example wanted to use it in one company that had poor infra and it didn't worked well with Jenskins for example, so I couldn't easily bypass it. 

Also Git LFS worked weird with rebases and people had a lot of trouble with it when jumping between branches if I remember correctly.

It works but experience is suboptimal.

**Comment by aslushnikov** [comment]
Hey folks! Here's an update on screenshots and blurring.

I see lots of you requested a "blur" option to pre-blur images before comparison. While I imagine it can help with certain
issues, it's a very big hammer, so I wonder if we can do a more delicate job.

### I'd appreciate if you could **share screenshots (actual / expected) that fail for you with regular diff, but pass with preblur**. This way we'll have some real-world data to play with!

**Comment by aslushnikov** [comment]
Many folks mentioned that they want pre-blur to avoid snapshot failures due to a few pixel differences.

For them, a new options [has landed](https://github.com/microsoft/playwright/pull/12169) on tip-of-tree: `pixelCount` and `pixelRatio`. Please give them a try and let me know, if you still need preblur! 

```bash
$ npm i @playwright/test@next
```

**Comment by shamrin** [comment]
Thank you for improving visual regression features, @aslushnikov!

You may find the implementation experience of gemini-testing project useful. Some pointers:

* [CIEDE2000 tolerance measure](https://github.com/gemini-testing/hermione#tolerance)
* [ignoring (font?) anti-aliasing](https://github.com/gemini-testing/looks-same#comparing-images-with-ignoring-antialiasing), based on [research paper](http://www.eejournal.ktu.lt/index.php/elt/article/view/10058/5000)
* [ignoring text input caret](https://github.com/gemini-testing/looks-same#comparing-images-with-ignoring-caret)

Several years ago we've great success using [Gemini](https://github.com/gemini-testing/gemini) for visual regression testing. We used Gemini built-in web UI (either [Gemini GUI](https://github.com/gemini-testing/gemini-gui) or [html-reporter](https://github.com/gemini-testing/html-reporter) - don't remember which) to choose changed images worth committing to Git. And during PR review we used built-in GitHub image diff. We had with very few false positives in image diffing. Unfortunately, false positives rate was not zero - mostly due to subtle browser timing/random fluctuations.

Gemini is deprecated now, replaced by [Hermione](https://github.com/gemini-testing/hermione), from the same authors. I haven't used it, but it seems to use the same approach for image diffing. The core is in [`looks-same`](https://github.com/gemini-testing/looks-same) and [gemini-core](https://github.com/gemini-testing/gemini-core#compare) libraries.

**Comment by aslushnikov** [comment]
Thanks @shamrin for the pointers! I'll read your links in more details later to get a better understanding, but so far we already do most of the things:

- instead of using [CIEDE2000](https://github.com/gemini-testing/hermione#tolerance), `pixelmatch` uses color difference in YIQ color space
- `pixelmatch` uses the same algorithm based on the same whitepaper to ignore anti-aliasing
- we hide text input caret on the browser level before making a screenshot

**Comment by ayroblu** [comment]
Hey! @aslushnikov I updated @florianbepunkt's original port of jest-image-snapshot to playwright test runner here: https://github.com/ayroblu/playwright-image-snapshot. Basically it looks VERY similar to playwright's existing [golden.ts](https://github.com/microsoft/playwright/blob/086333cd60625fcac0527c4ebfcef003a3be7183/packages/playwright-test/src/matchers/golden.ts) compare api and as you can see in [matcher.ts](https://github.com/ayroblu/playwright-image-snapshot/blob/main/src/matcher.ts#L50).

The main benefit it is that it uses SSIM. I also updated how the diff is done so it's similar to pixelmatch's greyscale background which is super useful.

<img width="1019" alt="image" src="https://user-images.githubusercontent.com/4915682/154841893-4db53094-7109-4371-85f7-676ceffd690f.png">

```ts
    expect(await page.screenshot()).toMatchImageSnapshot(test.info(), [
      name,
      "1-initial-load.png",
    ]);
```

Would love to have this SSIM option ported to playwright test as `TestInfo` is not exposed implicitly which makes the api usage a bit ugly.
I'm also hoping not to need to supply a file name by default, seems unnecessary.

**Comment by aslushnikov** [comment]
For the record: docker integration depends on global fixtures, so moving them forward.

**Comment by aslushnikov** [comment]
@bezyakina not sure for 1.21 (we're about to finalize this version), but still possible! It all depends on how much our users need it.

So could you please file this separately to our bug tracker as a feature request? The more likes / upvotes it will collect, the higher priority will be for us, and the faster we'll implement it!

**Comment by AllanMedeiros** [comment]
Hey there!
Not sure if would be better to open another feature request, but https://github.com/jz-jess/RobotEyes has an interesting feature to ignore an array of UI elements in the image comparison, as these elements will be blurred, helping to achieve a higher percentage of fidelity (+95%) comparison. RobotEyes uses Imagemagick in the background which is a really powerful tool for image comparison.
The idea is to ignore data elements from the screen before comparison is done. Taking that into account would require to set a different tolerance for each web page in the application, as each one can have different amount of UI elements with data.
I've seen comments about blur, but it doesn't seem to be related to this...
Thank you.

**Comment by aslushnikov** [comment]
@AllanMedeiros you can use the [`mask`](https://playwright.dev/docs/api/class-page#page-screenshot-option-mask) api to mask elements on the screenshot. This should help!

**Comment by leotg130** [comment]
Would it be possible to do visual diff's even when the snapshot sizes differ (`Sizes differ; expected image...`)? Right now it seems the Playwright VRT refuses to do visual diffs for such snapshots.

Storybook addon storyshots at least has this feature, though it might come from pixelmatch, not sure. It's very convenient, as often there's white space changes and you can easily see that some padding has appeared somewhere.

**Comment by tovab** [comment]
Can you add an option for ignoring diffs where there is just some slight shift in location of pixels?
I don't want to use threshold or maxDiffPixels for this because those options would cause false positives, i.e. they would cause the tests to ignore actual regressions.
Here is an example of a diff that I would like to ignore:
![reactions-selector-actual](https://user-images.githubusercontent.com/8853082/168743257-a38b69ae-4eec-4a4c-8f6b-d85c36e6049d.jpeg)
![reactions-selector-diff](https://user-images.githubusercontent.com/8853082/168743262-4c3ba521-f764-4a2a-8399-377918029981.jpeg)
![reactions-selector-expected](https://user-images.githubusercontent.com/8853082/168743265-84816ee9-d2e7-4a47-8fd6-b73aaff64af4.jpeg)

**Comment by KirProkopchik** [comment]
Hi @aslushnikov 
For now playwright can not compare the reference screenshot with actual instance in case if they have not equal resolution. 
This is not a big problem. But when I try to update such screenshot I also faced the error message about resolution mismatch. 
`Error: Image sizes do not match.
`
It is necessary to find such a screenshot and delete it. Then run the update again. This is not very convenient, for example, after updating the browser version. When many screenshots can be rendered differently.

**Comment by emilio-martinez** [comment]
@KirProkopchik different browsers and machines will produce different results. Have you tried running e2e tests within a [docker image](https://playwright.dev/docs/docker)? We've been doing this for a few months and the results have been quite stable

**Comment by sean-perkins** [comment]
I agree that adding support for blurring would be extremely helpful. We are making use of `maxDiffPixelRatio`, but it is much of a guessing game of what value is appropriate to counter antialiasing/rendering differences vs. actual issues. 

For example, I am currently having to set the value to `0.0005` to catch issues like this one (a real issue):

<details>
<img src="https://user-images.githubusercontent.com/13732623/169148787-40a199e2-e2c0-4650-bb7a-a5e708d67c8a.png" />
</details>

While wanting to skip/ignore false positives like this:
<details>
<img src="https://user-images.githubusercontent.com/13732623/169149048-b6333961-855d-497e-be89-46fa23a2ab09.png" />
</details>

I worry with just the arbitrary pixel ratio of pixel differences, we will be adjusting that value overtime as certain screenshots will be more impacted by antialiasing and rendering imperfections. 

We use Playwright for [Ionic Framework](https://github.com/ionic-team/ionic-framework), and test against Chrome/Firefox/Safari to verify correct Android and iOS design implementations for the web components.

**Comment by kenkku** [comment]
We have an interesting problem that's probably a common case when doing visual regression testing: we're taking a screenshot of an element (selected with a locator) that has a non-integer height. This results in an interesting problem where (at least when the device pixel ratio is 1) depending on what is on the rest of the page, sometimes the screenshot has a different height, or includes one extra row of the background color outside the element. I think this could even happen for elements that have an integer height, but that are positioned around elements that don't.

I think the most useful case for visual regression testing is for individual elements, and this does make it very hard to test those if you have any sub-pixel heights (or widths) on the page.

**Comment by tovab** [comment]
> We have an interesting problem that's probably a common case when doing visual regression testing: we're taking a screenshot of an element (selected with a locator) ....
> I think the most useful case for visual regression testing is for individual elements, and this does make it very hard to test those if you have any sub-pixel heights (or widths) on the page.

I experienced this issue and my workaround was to take an image of the entire page and then crop the image to select the desired element before comparing. 

`const dimensions = await element.boundingBox();
      expect(await page.screenshot({
        type: 'jpeg',
        clip: dimensions as {x; y; width; height },
      })).toMatchSnapshot(`${name}.jpeg`, {});`

**Comment by nmiddendorff** [comment]
With the release of [1.21](https://github.com/microsoft/playwright/releases/tag/v1.21.0), Playwright now has the "Slider Diff View" which is great for comparing visual changes on the `.toMatchSnapshot()` assertion.

I'm curious how others plan to incorporate this into their software development workflow! It seems like the biggest piece missing from Playwright now is the ability to approve changes outside of running the app locally. (This is where backstopjs is still useful). Have others come up with a way to create some type of workflow on the PR that allows teams to easily review and approve changes?

To be clear, I'm not necessarily saying that this **should be** part of Playwright.

![image](https://user-images.githubusercontent.com/8575456/179271205-991452ec-f0b5-4bbb-8b1e-d3ad86fb9280.png)

**Comment by anduingaiden** [comment]
Is there any way to take one screenshot of the entire page?
We have many cases of "long pages" and for now, we are instructing our test script to, multiple times, scroll down the page and take a snapshot until we reach the end of the page. Than, we compare all the screenshots.

Is there an easier way to do it? If not, do u intend to add this kind of option? 

Thanks in advance!

**Comment by ayroblu** [comment]
@anduingaiden Please see: https://playwright.dev/docs/screenshots#full-page-screenshots

**Comment by anduingaiden** [comment]
> @anduingaiden Please see: https://playwright.dev/docs/screenshots#full-page-screenshots

It's clear that I did not explore this part of the documentation yet. Sorry about that and thank you very much.

**Comment by pastelsky** [comment]
Another example of where text rendering causes flake and blurring could've helped —
![image](https://user-images.githubusercontent.com/8946207/201529682-269b3b33-91b1-4922-82d9-8b46436e803e.png)

**Comment by mrmckeb** [comment]
Hi @pastelsky, are those screenshots taken on different operating systems? In a like-for-like environment, you shouldn't see flake from text rendering differences.

Different operating systems can have different default fonts, and (which appears to be the case here) different text rendering approaches.

**Comment by aslushnikov** [comment]
Hey folks! if you have examples of **PNG screenshots** that are taken on the **same browser** and **same OS**
yet are different due to anti-aliasing issues, could you please attach the "expected", "actual" and "diff" images here?

This information will help with my experiments with fighting browser rendering non-determinism.

**Comment by EchelonFour** [comment]
> I'm curious how others plan to incorporate this into their software development workflow! It seems like the biggest piece missing from Playwright now is the ability to approve changes outside of running the app locally. (This is where backstopjs is still useful). Have others come up with a way to create some type of workflow on the PR that allows teams to easily review and approve changes?

I have come up with a system where the devs can comment on the PR to run a CI task that reruns the tests with an `--update-snapshots` flag and pushes the any changes to the PR branch. But this requires rerunning the entire tests again, which is pretty slow considering the test report from the last run already has the accepted new snapshots in it.

It would be nice to have some kind of "accept snapshots" command we could run that takes an output from a test run where the snapshots comparison failed and it updates them from that. Even if it needs some kind of special report format, that would speed up this part of the workflow considerably.

**Comment by gselsidi** [comment]
These are all inside Docker the only difference is Docker on OS X Mac M1 and Docker on Ubuntu Latest

![member-profile-preview-3-actual](https://user-images.githubusercontent.com/7006827/204597012-b4f65fb7-0d5e-4220-9fa8-351a4abeadb1.png)
![member-profile-preview-3-diff](https://user-images.githubusercontent.com/7006827/204597016-6c12aaf7-095f-4ec9-b82a-998e711f44c6.png)
![member-profile-preview-3-expected](https://user-images.githubusercontent.com/7006827/204597017-0d230e9a-fa40-4934-9fd7-b2be53d928f5.png)

**Comment by aslushnikov** [comment]
@gselsidi thank you for the sample!

**Comment by gselsidi** [comment]
I'll try to get some more as they come along, but i noticed the above occurs when taking snapshots of individual elements as opposed to the whole page. The whole page I'm able to use .0001 pixel ratio.

**Comment by gselsidi** [comment]
linking this here incase it applies:

https://github.com/microsoft/playwright/issues/19417

**Comment by nikicat** [comment]
I had different screenshots with antialiased fonts between my ArchLinux laptop and Ubuntu 20.04 in Docker (it's used by default by GitHub Actions). The following chromium flags helped me to get identical screenshots:
```
--font-render-hinting=none
--disable-skia-runtime-opts
--disable-font-subpixel-positioning
--disable-lcd-text
```

**Comment by phungleson** [comment]
We have similar issues with webkit on mac around emojis, I am not sure if we can provide further information to make debugging/fixing the issue easier?

It looks like [mask](https://playwright.dev/docs/api/class-page#page-screenshot-option-mask) is not available to configure at `PlaywrightTestConfig` level?

**Comment by thekp** [comment]
>  I had different screenshots with antialiased fonts between my ArchLinux laptop and Ubuntu 20.04 in Docker (it's used by default by GitHub Actions). The following Chromium flags helped me to get identical screenshots:
> --font-render-hinting=none
> --disable-skia-runtime-opts
> --disable-font-subpixel-positioning
> --disable-lcd-text

@nikicat Where exactly did you add these flags in your playwright project?

**Comment by nikicat** [comment]
@thekp I pass them to `playwright.chromium.launch(args=[...])` [here](https://github.com/Donate4Fun/donate4fun/blob/master/donate4fun/screenshot.py#L46)
(`chromium_flags()` function [is overriden inside a test](https://github.com/Donate4Fun/donate4fun/blob/master/tests/test_web.py#L95-L102)).

**Comment by ayroblu** [comment]
One of the things we noticed is that our focus is somewhat different between runs (imagine loading a page with a text input). Wondering if there's anything we can do to improve reliability apart from just manually blurring and focusing

**Comment by matthias-ccri** [comment]
My discrepancy was resolved by passing the `--disable-remote-fonts` flag to chromium.
```
    projects: [
        {
            name: 'chromium',
            use: {
                ...devices['Desktop Chrome'],
                launchOptions: {
                    args: [
                        '--disable-remote-fonts',
                    ],
                },
            },
        },
    ],
```

**Comment by mscottford** [comment]
I wonder if some of the font rendering discrepancies might be because of local fonts being used instead of web fonts. For example, I have the Rambla font installed locally on my Mac, but my site also pulls that font in via CSS. In that case, consistently running the tests in an environment that does not have those fonts installed locally might address the problem. This might mean replacing the "expected" image with one from an environment that doesn't have the font installed.

M. Scott Ford
Co-Founder & Chief Code Whisperer (CTO)
Corgibytes, LLC
804.596.2375 x701
pronouns: he/him
***@***.***
https://corgibytes.com ( https://corgibytes.com/ )

Have you read our First Round Review ( http://firstround.com/review/forget-technical-debt-heres-how-to-build-technical-wealth/ ) article about paying off technical debt?

Love refactoring and TDD? Join us at LegacyCode.Rocks ( http://LegacyCode.Rocks ) for virtual meetups, podcasts, and more.

Sent via Superhuman ( ***@***.*** )

**Comment by deviantintegral** [comment]
I've noticed an issue with webkit image rendering where it doesn't seem to be consistent. Look at the image of the flowers in this picture:
![a-basic-page-with-embed-images-ID-2008-1-expected](https://github.com/microsoft/playwright/assets/255023/0c0b38d0-71e2-455e-ae9b-126b1a5214d5)

And this one:

![a-basic-page-with-embed-images-ID-2008-1-actual](https://github.com/microsoft/playwright/assets/255023/2d7a576d-fa65-4fe2-9fd8-3aa16b6148c5)

There's exactly 30 pixels different - and what's interesting is that when it fails, it's _always_ 30 pixels.

![a-basic-page-with-embed-images-ID-2008-1-diff](https://github.com/microsoft/playwright/assets/255023/b191abcd-744a-4fcb-b7b2-2f20889a0016)

If you flip between the two images, one of them appears more aliased or slightly blurred or something. The image is a lossy webp image, so I suppose it could be rendering the image isn't consistent?

Anyone know if this is expected - something like webkit rendering the image in stages? We're already waiting on the `complete` property so JavaScript and playwright consider the image loaded.

**Comment by gselsidi** [comment]
> I've noticed an issue with webkit image rendering where it doesn't seem to be consistent. Look at the image of the flowers in this picture: ![a-basic-page-with-embed-images-ID-2008-1-expected](https://user-images.githubusercontent.com/255023/247762601-0c0b38d0-71e2-455e-ae9b-126b1a5214d5.png)
> 
> And this one:
> 
> ![a-basic-page-with-embed-images-ID-2008-1-actual](https://user-images.githubusercontent.com/255023/247762672-2d7a576d-fa65-4fe2-9fd8-3aa16b6148c5.png)
> 
> There's exactly 30 pixels different - and what's interesting is that when it fails, it's _always_ 30 pixels.
> 
> ![a-basic-page-with-embed-images-ID-2008-1-diff](https://user-images.githubusercontent.com/255023/247762780-b191abcd-744a-4fcb-b7b2-2f20889a0016.png)
> 
> If you flip between the two images, one of them appears more aliased or slightly blurred or something. The image is a lossy webp image, so I suppose it could be rendering the image isn't consistent?
> 
> Anyone know if this is expected - something like webkit rendering the image in stages? We're already waiting on the `complete` property so JavaScript and playwright consider the image loaded.

There are a few tickets with all the visual stuff, if I remember correctly there is an issue with WebKit rendering outside of playwrights control. 

At the end I just gave up at 100% pixel perfection and allowed a % of variance. 

Even 95% accuracy is still a feat on its own, in reality you’ll probably get to 99.9% good enough.

but yeah would be cool to get 100% so we know if things degrade in the future we degrade from 100% as opposed to a starting point of 95%

also make sure you use docker if not always running against the actual same physical machine

**Comment by deviantintegral** [comment]
> There are a few tickets with all the visual stuff, if I remember correctly there is an issue with WebKit rendering outside of playwrights control.
> At the end I just gave up at 100% pixel perfection and allowed a % of variance.

_nods_ yeah, that's what I figured. I'm currently working around it by setting `maxDiffPixels` if the `browserName` is webkit. Hopefully we can maintain 100% pixel coverage in Chrome or Firefox.

> also make sure you use docker if not always running against the actual same physical machine

Good reminder. We're doing that with https://github.com/deviantintegral/ddev-playwright and the above screenshots are from running tests in a loop until they fail, all in the same environment.

**Comment by pastelsky** [comment]
There's a separate issue regarding adding support for docker at https://github.com/microsoft/playwright/issues/20954 so that visual regression tests can run in a consistent environment and environment-related differences are negated.  

It would be helpful to receive upvotes there from folks here if that's something you need.

**Comment by mfucci-medable** [comment]
I am encountering the same issue with chromium (and webkit at an even higher frequency, too high so we disabled it).

**Version**: Playwright 1.38.1 (but the issue is reproducible as well in 1.39.0)
**Env**: running in ubuntu:jammy on an Apple M1 Pro (but the issue happens in our Linux CI pipeline as well, running in docker makes it pixel perfect between local and CI)
**What happens**: About 5% of the time, randomly one letter is incorrectly positionned, always the same letter. On other screenshots, it might 2 -3 letters, sometime in the middle of a word.
**More info:**

- No network call, the css is inlined before the HTML.
- Using chromium arguments (no improvements before / after enabling those arguments):

              '--font-render-hinting=none',
              '--disable-skia-runtime-opts',
              '--disable-system-font-check',
              '--disable-font-subpixel-positioning',
              '--disable-lcd-text',
              '--disable-remote-fonts',

**My guess**: this issue never happens on other screenshots that we are taking using exactly the same configuration, so it has to do with something in the HTML / CSS (that I am probably not allowed to share here)...

Actual / expected / diff (triggering here on the pseudo-locale test but might happen as well on the en-US version):

![actual](https://github.com/microsoft/playwright/assets/148731315/f9e8cf77-fa99-4d26-a898-956d44325949)
![expected](https://github.com/microsoft/playwright/assets/148731315/3e04d9da-1b34-4d3b-8656-bca544a3b8fe)
![diff](https://github.com/microsoft/playwright/assets/148731315/4978014a-edf4-4897-a32a-481ad029a6f1)

**Comment by viktor-avd** [comment]
> ## From maintainers
> Hey folks! if you have examples of **PNG screenshots** that are taken on the **same browser** and **same OS** yet are different due to anti-aliasing issues, could you please attach the "expected", "actual" and "diff" images here?
> 
> This information will help with our experiments with fighting browser rendering non-determinism.

Hi, this appeared with the latest version, nothing like this happened before with the same code and configuration.

Playwright version: **1.41.1**
Docker image: **mcr.microsoft.com/playwright:v1.41.1-jammy**
Chrome without args / the same with next args:

```
'--disable-skia-runtime-opts',
'--disable-system-font-check',
'--disable-remote-fonts',
'--font-render-hinting=none',
'--disable-font-subpixel-positioning',
'--disable-lcd-text'
```

**Example 1:**

_expected_
![expected](https://github.com/microsoft/playwright/assets/93348096/cc09a090-b4a1-49ce-b1fe-9e795e67285c)

_actual_
![actual](https://github.com/microsoft/playwright/assets/93348096/50abcb0e-80c4-4263-81f4-d9ff1e9556ea)

_diff_
![diff](https://github.com/microsoft/playwright/assets/93348096/00764238-77bb-4dfe-a7fa-8a041bcd3883)


**Example 2:**

_expected_
![actual-rule](https://github.com/microsoft/playwright/assets/93348096/50da82b0-14ec-4bed-894c-c164331b8ff5)

_actual_
![expected-rule](https://github.com/microsoft/playwright/assets/93348096/78874329-bff0-4e50-9211-64380231ff05)

_diff_
![diff-rule](https://github.com/microsoft/playwright/assets/93348096/2ca6469b-1ac3-40de-b1e5-3300c879079a)

**Comment by deviantintegral** [comment]
An update from our experiences above: We found that increasing `maxDiffPixels` (or `maxDiffPixelRatio`) to a level that could avoid false failures also led to too many regressions slipping through visual comparisons. However, the `threshold` option as documented https://playwright.dev/docs/api/class-pageassertions#page-assertions-to-have-screenshot-2 worked for us. Once we increased that from the default `0.2` to `0.3`, we've had no false failures _or_ missed regressions.

**Comment by ferrata** [comment]
Hey folks! Any updates on this feature? 😅

I am facing the same behavior after switching to a recent Playwright version (`1.49.0`).

<details><summary>Diff</summary>
<img src="https://github.com/user-attachments/assets/3b5e57aa-b876-43db-8eb0-32ee5c398f39" />
</details> 

<details><summary>Actual</summary>
<img src="https://github.com/user-attachments/assets/a494f55b-7d6b-42a4-9fde-7a537d9b166d" />
</details> 

<details><summary>Expected</summary>
<img src="https://github.com/user-attachments/assets/f2255aa4-2f88-4492-a039-2eed0162470d" />
</details> 

FWIW, it was working correctly on the Playwright version `1.41.2` with the `PLAYWRIGHT_CHROMIUM_USE_HEADLESS_NEW` variable on.

Thanks!

**Comment by AntonioVentilii** [comment]
Hi all!

I am having the same issue but in Safari device and with an emoji:

```json
	{
		name: 'Safari',
		use: devices['Desktop Safari']
	},
```



![Image](https://github.com/user-attachments/assets/67a9983b-fccd-4ca0-be2c-7cb5ef2cc6bd)
![Image](https://github.com/user-attachments/assets/1319a855-0298-4909-b5a8-cba5dc248d61)

**Comment by raythurnvoid** [comment]
In the CI (GitHub Actions), same runner, same configuration, same tests same versions.

We update the screenshots in the CI, we re-run the tests and we get small screenshots diffs on texts.

**Comment by Someraj** [comment]
In CI, the slider does not work if we are merging from blob reporter to html. The slider in left blank in the html report.

**Comment by yairEO** [comment]
I am seeing the exact same problem as @raythurnvoid describes. Always the same diff on the same pixels and only in the CI, using `mcr.microsoft.com/playwright:v1.53.0-jammy` and `"@playwright/test": "^1.53.0"`

Tests run only in `Chrome`, and for a single viewport size.

Some very specific screenshots are ALWAYS flaky on certain parts. 

This is affects out team severely as many PRs are blocked from merging, and developers are not happy having to run retries over and over until the flakiness is randomly resolved..

I don't know the root cause of this - if it's a playwright issue, a docker image issue or a CSS issue.. extremely difficult to work like this, as we have tons of screenshots tests

**Comment by deviantintegral** [comment]
We recently fixed a problem where elements on a page were being rendered with fractional pixels, even at the largest viewport. In particular, anything rendered with half a pixel was flaky. We ended up fixing the CSS. The easiest way to identify it is to inspect elements starting at the point, or just above. where the image comparison starts to show differences.

**Comment by yairEO** [comment]
@deviantintegral - what do you mean "half a pixel"? It's extremely common in CSS to work in non-px units, and half-pixels are extremely common. There is no fix for this, since it is everywhere...

**Comment by deviantintegral** [comment]
I mean inspecting elements in browser dev tools would show something like a height of "200.5" pixels.

My understanding was that browsers don't render 0.5 pixels consistently... but I'm also not a front-end or browser expert.

**Comment by yairEO** [comment]
that is normal in all web content to measure elements` sized with half pixels. 

<img width="356" alt="Image" src="https://github.com/user-attachments/assets/de02e458-88e0-4388-812f-4e07dcad554c" />

Elements dimensions are fluid by nature. This shouldn't be an issue.. or else screenshot-based testing would have never worked..

**Comment by ivanalemunioz** [comment]
Hi, everyone. I’ve been working on a solution to visually debug my browser automation crashes, and I’ve now made it free for the public to use: https://buglesstack.com

I don’t mean to spam or bother you; I just want to share this solution. If you have any questions or suggestions for improvement, feel free to contact me.

Thanks for reading! ❤️

**Comment by jack-bradshaw** [comment]
I've been having trouble on image heavy pages with Playwright Java, and found that disabling a bunch of browser features was necessary to eliminate flakiness. Right now I'm using:

BrowserType.LaunchOptions()
        .setHeadless(true)
        .setArgs(listOf(
            "--disable-gpu",
            "--disable-font-subpixel-positioning",
            "--disable-lcd-text",
            "--disable-threaded-animation",
            "--disable-threaded-scrolling",
            "--disable-in-process-stack-traces",
            "--disable-checker-imaging",
            "--force-color-profile=srgb"
        ))
    browser = playwright.chromium().launch(launchOptions)

These basically disable animations, scrolling, GPU acceleration, and anything else that might subtly adjust the rendered pixels.

**Comment by prakersh** [comment]
For marking specific UI elements during visual comparison, [onUI](https://github.com/onllm-dev/onUI) might be a useful companion tool — it's a free, open-source browser extension that provides an annotation overlay to label and highlight specific elements on any webpage. The annotations include element coordinates and metadata that could help identify what changed between snapshots.
